Update Integration Tests

eerhardt opened the pull request:
1. Add retry / resilience logic to the services that have slow starting containers (MySQL, Oracle, and SqlServer). Each of these containers can take more than 30 seconds to start on my machine.
2. Remove unnecessary containers from the IntegrationServiceA project. Adding double the containers make the startup take even longer. So just use AddXYX resources and remove AddXYZContainer services.
3. A couple other cosmetic changes - Use "rabbitmq" vs. "rabbit" - Log the responseContent in the SqlServer test


This won't be the last refactoring to these tests, but I think it is a step in the right direction.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1540)

eerhardt committed:
Update Integration Tests

1. Add retry / resilience logic to the services that have slow starting containers (MySQL, Oracle, and SqlServer). Each of these containers can take more than 30 seconds to start on my machine.
2. Remove unnecessary containers from the IntegrationServiceA project. Adding double the containers make the startup take even longer. So just use AddXYX resources and remove AddXYZContainer services.
3. A couple other cosmetic changes
    - Use "rabbitmq" vs. "rabbit"
    - Log the responseContent in the SqlServer test

mitchdenny commented:
How do we ensure that the AddXyzContainer resources don't bit rot

eerhardt commented:
> How do we ensure that the AddXyzContainer resources don't bit rot

Other tests.

These tests weren't ensuring those resources don't bit rot - because they weren't being used.

eerhardt commented:
Maybe one thing we could do next is add another bool to CreateTestProgram that would switch between `AddXYZ` and `AddXYZContainer`. Then the same tests could run against both. But at any given time, only 1 set of resources would be created - either `AddXYZ` or `AddXYZContainer`.

Thoughts?

davidfowl commented on tests/testproject/TestProject.IntegrationServiceA/SqlServer/SqlServerExtensions.cs:
```diff
@@ -14,13 +15,21 @@ private static async Task<IResult> VerifySqlServerAsync(SqlConnection connection
         try
         {
-            await connection.OpenAsync();
+            var policy = Policy
```
Why does this need retries?

eerhardt replied:
Because the SqlServer container takes like ~30-40 seconds to start up. If you run this test by itself (or it is the first one being executed), the server isn't available and the test fails.

mitchdenny commented:
> Other tests.

I looked through some of the resource specific functional tests and they seem to rely on the TestProgram as well to spin up the containers. With this change none of them will be testing `AddXYZContainer(...)` variants.

What the top level test relied upon (incorrectly in some instances) was the fact that we had healthchecks registered, so by hitting the healthcheck endpoint we could verify that the integrations were working (some components no-op healthchecks without additional config).

mitchdenny commented:
We need to either break up the integration test so they are smaller and focused on smaller groups of components or we need to read the AddXYZContainer extensions.

If we don't I don't think we've got an integration test that executes that code path.

I think with a larger timeout and improved retry logic it should be okay. And if there is instability starting larger groups of services we want to know about it.

In time I suspect that we'll end up with multiple integration test projects.

eerhardt committed:
Add more time for Cosmos, since the emulator can take over a minute to start.

Remove retries in Cosmos test. We don't need to retry at the test side, retries should be done inside the integration app.

eerhardt commented:
@mitchdenny - can I get your eyes on this? The tests each pass individually on my machine.

My next step will be to split the Aspire.Hosting.Tests apart. We will leave the "unit tests" for Aspire.Hosting in Aspire.Hosting.Tests. And these integration tests will move to a new "EndToEnd" functional tests by themselves.

radical commented on tests/Aspire.Hosting.Tests/Cosmos/CosmosFunctionalTests.cs:
```diff
@@ -26,15 +24,9 @@ public async Task VerifyCosmosWorks()
 
         using var cts = new CancellationTokenSource(TimeSpan.FromMinutes(6));
 
-        await RetryPolicy.Handle<HttpRequestException>()
```
Just curious - why remove this? Is there a high cost to having this when the service starts up fast enough?

radical replied:
> Retrying on any HttpRequestException means when this test is actually broken the test will retry and retry over and over until the timeout expires.

That makes sense.

> The retry/waiting for the Cosmos emulator is already happening in the Cosmos client. It will retry until the server comes up.

Ok, for other cases like mysql, and for cosmos (real) the user would be responsible for adding such retries? Do they have a way of differentiating and retrying only for "connection errors", and not *any* `HttpRequestException`?

eerhardt replied:
> for other cases like mysql, and for cosmos (real) the user would be responsible for adding such retries?

it all depends on the client you are using. StackExchange.Redis has an option for it to do the retries, so we enable that option by default:
https://github.com/dotnet/aspire/blob/5043b31f5e4a50cf489ccb60c129d3edb1b42c0b/src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs#L201-L202

RabbitMQ doesn't, so we have to write it ourselves:

https://github.com/dotnet/aspire/blob/5043b31f5e4a50cf489ccb60c129d3edb1b42c0b/src/Components/Aspire.RabbitMQ.Client/AspireRabbitMQExtensions.cs#L154-L180

> Do they have a way of differentiating and retrying only for "connection errors", and not any HttpRequestException?

Yes, Polly has the feature that you can supply a predicate and inspect the Exception to check for what kind of failure - and only retry on certain exceptions.

radical replied:
Thank you for the responses!

radical commented on tests/testproject/TestProject.AppHost/TestProgram.cs:
```diff
@@ -38,50 +38,32 @@ private TestProgram(string[] args, Assembly assembly, bool includeIntegrationSer
                 .AddDatabase(postgresDbName);
-            var rabbitmqContainer = AppBuilder.AddRabbitMQContainer("rabbitmqcontainer");
-            var mongodbContainer = AppBuilder.AddMongoDBContainer("mongodbcontainer")
+            var rabbitmq = AppBuilder.AddRabbitMQContainer("rabbitmq");
+            var mongodb = AppBuilder.AddMongoDBContainer("mongodb")
```
Did you mean to use the non-container versions of these?

Also, https://github.com/dotnet/aspire/blob/5043b31f5e4a50cf489ccb60c129d3edb1b42c0b/src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs#L39-L47

.. the code comment says that `AddMongoDb` will use a container for local development. Would it still be using a container then?

(unrelated: how does it decide that this is "local development"?)

mitchdenny replied:
This is reparation for a change we have coming in P4 where we are consolidating the XYZResource and XYZContainerResource types and extension methods. They both work for local development, the difference is intent when it comes deployment time. But in P4 we'll be introucing another mechanism to specify that intent.

eerhardt replied:
I've updated these to use the non-container versions. Nice catch.

mitchdenny commented on tests/testproject/TestProject.IntegrationServiceA/MySql/MySqlExtensions.cs:
```diff
@@ -14,13 +15,21 @@ private static async Task<IResult> VerifyMySqlAsync(MySqlConnection connection)
-            var results = await command.ExecuteReaderAsync();
+            return await policy.ExecuteAsync(async () =>
+            {
+                await connection.OpenAsync();
```
This may not be a blocker given this is a test case. But if `OpenAsync(...)` succeeds but the query fails (for whatever transient reason) then it might just continue to fail because it'll start failing on subseuqnet `OpenAsync(...)` calls.

eerhardt replied:
I changed this to only retry on connection open.

eerhardt committed:
PR feedback

Use non-Container overloads
Only retry on connect.
